doc/api/esm.md

```diff
@@ -658,150 +675,105 @@ export async function resolve(specifier, context, defaultResolve) {
 }
 ```
 
-#### `getFormat(url, context, defaultGetFormat)`
+#### `load(url, context, defaultLoad)`
 
 > Note: The loaders API is being redesigned. This hook may disappear or its
 > signature may change. Do not rely on the API described below.
 
+> Note: In a previous version of this API, this was split across 3 separate, now
+> deprecated, hooks (`getFormat`, `getSource`, and `transformSource`).
+
 * `url` {string}
 * `context` {Object}
-* `defaultGetFormat` {Function}
+  * `format` {string|null|undefined} The format optionally supplied by the
+  `resolve` hook.
+* `defaultLoad` {Function}
 * Returns: {Object}
   * `format` {string}
+  * `source` {string|ArrayBuffer|TypedArray}
 
-The `getFormat` hook provides a way to define a custom method of determining how
-a URL should be interpreted. The `format` returned also affects what the
-acceptable forms of source values are for a module when parsing. This can be one
-of the following:
+The `load` hook provides a way to define a custom method of determining how
+a URL should be interpreted, retrieved, and parsed.
 
-| `format`     | Description                    | Acceptable Types For `source` Returned by `getSource` or `transformSource` |
+The final value of `format` must be one of the following:
+
+| `format`     | Description                    | Acceptable Types For `source` Returned by `resolve` or `load` |
 | ------------ | ------------------------------ | -------------------------------------------------------------------------- |
 | `'builtin'`  | Load a Node.js builtin module  | Not applicable                                                             |
 | `'commonjs'` | Load a Node.js CommonJS module | Not applicable                                                             |
 | `'json'`     | Load a JSON file               | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][] }                      |
 | `'module'`   | Load an ES module              | { [`string`][], [`ArrayBuffer`][], [`TypedArray`][] }                      |
 | `'wasm'`     | Load a WebAssembly module      | { [`ArrayBuffer`][], [`TypedArray`][] }                                    |
 
-Note: These types all correspond to classes defined in ECMAScript.
-
-* The specific [`ArrayBuffer`][] object is a [`SharedArrayBuffer`][].
-* The specific [`TypedArray`][] object is a [`Uint8Array`][].
-
-Note: If the source value of a text-based format (i.e., `'json'`, `'module'`) is
-not a string, it is converted to a string using [`util.TextDecoder`][].
+> Note: When format is `'commonjs'`, the value of `source` is ignored (due to
+> incompatibilities between internal commonjs and ES modules APIs). This may
+> change in the future.
```

Comment:
```suggestion
> Note: When format is `'commonjs'`, the value of `source` is ignored (due to
> incompatibilities between internal commonjs and ES modules APIs). This may
> change in the future.

The value of `source` is ignored for type `'builtin'` because currently it is
not possible to replace the value of a Node.js builtin (core) module.

The value of `source` is ignored for type `'commonjs'` because the CommonJS
module loader is synchronous and the ES module loader is asynchronous, and
this difference presents a challenge for ES module loader hooks to modify
the loading of CommonJS modules. This limitation might be overcome in the
future.
```

@JakobJingleheimer @guybedford is this more-detailed explanation correct? I feel like this is a little more important than a blockquote aside, as returning `{ format: 'commonjs', source }` isn’t so much of an edge case that few loader authors will bump into it; I already have in the CoffeeScript loader example.

I’m also wondering if the last sentence should be stronger, like “This limitation might be overcome in the future, however there is a good chance that it is insurmountable.”

Reply:
It's _also_ correct, but it's not the real reason. As I understand it (from a combination of looking at the code and hearing from people, albeit dissenters, involved), the issue is human. There have been multiple implementations that solve the issue but were rejected due to opinion, not technical limitation.

But yes, it seems to me this is more than an edge-case.

Reply:
I'm not sure I agree, there are technical problems with various ways to make code run w/ this. It isn't that we can't make *something* since in general we can write code that does something; it is a question about what the code does and if it acts in a reliable or surprising manner. The technical limitation is not having a way to integrate agains the API design of `require.extensions` in a clean way. Doing so requires completely re-architecting how loaders would work and prevents things like in-memory transforms since that API is filepath based and returns a function to execute not something that can be used to describe various different module types.

Reply:
Well, yes, but not insurmountable. What I meant was the blocker is people, not that it's logically to functionally impossible.

But, last meeting we discussed how ESMLoader hooks could effectively replace what exists, in which case, it would not need to integrate with it. If we can get that going, I think this is moot.

Reply:
Getting back to my suggestion: a) is what I wrote correct? and b) do you consider it an improvement? I found the original text a bit vague (and yes, I know I helped write it over Slack 😄) and I’m trying to forestall people opening issues about this, which I feel like they’d be more likely to the less we explained here.

Reply:
Hmm. I would say what you've suggested above is truthful but not accurate.

I would say something like:

> ESMLoader hooks cannot modify the return of a namespaced CommonJS module because the CJS API does not expose a mechanism that facilitates doing so. This may change in future, but is a significant undertaking with many hurdles, and other avenues under consideration may be taken instead.

Reply:
I don’t know what “namespaced CommonJS module” means, and if _I_ don’t, I feel like average users probably don’t either.

What’s inaccurate about what I wrote? And I think we want to separate out `builtin`, don’t we, since it has a separate reason why it can’t be overridden, doesn’t it?

```suggestion
> Note: When format is `'commonjs'`, the value of `source` is ignored (due to
> incompatibilities between internal commonjs and ES modules APIs). This may
> change in the future.

The value of `source` is ignored for type `'builtin'` because currently it is
not possible to replace the value of a Node.js builtin (core) module.

The value of `source` is ignored for type `'commonjs'` because the CommonJS
module loader does not provide a mechanism for the ES module loader to
override the CommonJS module return value. This limitation might be overcome
in the future.
```

Reply:
RE "namespaced CJS modules", I don't know what they're called, it's just what the existing ESM doc refers to them as: https://nodejs.org/api/esm.html#esm_commonjs_namespaces

The CJS note and the CJS text below it seem redundant to me. I would just leave off the note then if providing the more detailed explanation below.

I thought the note formatting would be a more visually striking callout, but I don't have much of an opinion on it.

Yes, sorry, I was not suggesting excluding the builtin bit (I agreed, so I didn't mention it in the revisited CJS wording to avoid visual clutter).

Reply:
Sorry, I meant to delete the aside. How’s this?


```suggestion
The value of `source` is ignored for type `'builtin'` because currently it is
not possible to replace the value of a Node.js builtin (core) module.

The value of `source` is ignored for type `'commonjs'` because the CommonJS
module loader does not provide a mechanism for the ES module loader to
override the CommonJS module return value. This limitation might be overcome
in the future.
```

And maybe we could link the “CommonJS module return value” text to https://nodejs.org/api/esm.html#esm_commonjs_namespaces?

I wouldn’t add any additional text about how we’re aware of the issue, working on it, etc. because then people will just open issues like “what’s the status on this? when will you be done?” and so on.

Reply:
LGTM!

Reply:
Should we open an issue in nodejs/loaders to track the cjs issue and how we might address it?

Reply:
We can open it, but I don't have a good solution at this time so it would likely be mostly describing the problem.

Reply:
> We can open it, but I don’t have a good solution at this time so it would likely be mostly describing the problem.

There’s value in simply describing the problem. I think it should probably be in the main `nodejs/node` repo though, because it’s essentially a bug report; and that can be the “primary” issue tracking it. Better that one of us opens it than some user who won’t explain the problem as well as we can. And if it’s in the main repo, that will hopefully prevent users from opening (as many) new issues, as some might find the existing “canonical” issue.

We could _also_ have an issue in the loaders repo if we want a place to discuss potential solutions etc.

Reply:
Yeah, I second the value in detailing the problem.

I was thinking putting it in nodejs/loaders would keep it easier to manage (plus I have powers there, so I can do things haha). Plus keep it out of the way.

I think, unfortunately, it's not likely to better avoid dupes. Google search or nothing 😓

I'm happy to defer to your wisdom though.

Reply:
I think it’s a specific issue with code that’s in the `nodejs/node` repo, so therefore it belongs in `nodejs/node`. Whereas most issues in the loaders repo are discussions about the future of the loaders effort, rather than specific bugs.

In general I know from the modules work there was a desire on the part of the TSC and other collaborators to try to keep stuff in `nodejs/node` that might be relevant to other collaborators, since unless you happen to follow `nodejs/loaders` you wouldn’t get notified about things in there. I think this falls into that category of something that is important enough that we probably want the broader Node team aware of it, especially if we go so far as to finally remove `require.extensions`.